From ce814f988f8839da60afea4e2b9cd66ae8cf8089 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:59 -0400
Subject: [ethernet] qlcnic: Fix update of ethtool stats

Message-id: <1408626422-3941-136-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88661
O-Subject: [PATCH RHEL 7.1 135/138] qlcnic: Fix update of ethtool stats.
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Rajesh Borundia <rajesh.borundia@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id a0eaf75c03712b491b7a840b5836c8f1e2a09277

o Aggregating tx stats in adapter variable was resulting in
  an increase in stats even after no traffic was run and
  user runs ifconfig/ethtool command.
o qlcnic_update_stats used to accumulate stats in adapter
  struct at each function call, instead accumulate tx stats
  in local variable and then assign it to adapter structure.

Reported-by: Holger Kiehl <holger.kiehl@dwd.de>
Signed-off-by: Rajesh Borundia <rajesh.borundia@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
index 1b7f3db..141f116 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
@@ -1290,17 +1290,25 @@ static u64 *qlcnic_fill_stats(u64 *data, void *stats, int type)
 
 void qlcnic_update_stats(struct qlcnic_adapter *adapter)
 {
+ struct qlcnic_tx_queue_stats tx_stats;
  struct qlcnic_host_tx_ring *tx_ring;
  int ring;
 
+ memset(&tx_stats, 0, sizeof(tx_stats));
  for (ring = 0; ring < adapter->drv_tx_rings; ring++) {
   tx_ring = &adapter->tx_ring[ring];
-  adapter->stats.xmit_on += tx_ring->tx_stats.xmit_on;
-  adapter->stats.xmit_off += tx_ring->tx_stats.xmit_off;
-  adapter->stats.xmitcalled += tx_ring->tx_stats.xmit_called;
-  adapter->stats.xmitfinished += tx_ring->tx_stats.xmit_finished;
-  adapter->stats.txbytes += tx_ring->tx_stats.tx_bytes;
+  tx_stats.xmit_on += tx_ring->tx_stats.xmit_on;
+  tx_stats.xmit_off += tx_ring->tx_stats.xmit_off;
+  tx_stats.xmit_called += tx_ring->tx_stats.xmit_called;
+  tx_stats.xmit_finished += tx_ring->tx_stats.xmit_finished;
+  tx_stats.tx_bytes += tx_ring->tx_stats.tx_bytes;
  }
+
+ adapter->stats.xmit_on = tx_stats.xmit_on;
+ adapter->stats.xmit_off = tx_stats.xmit_off;
+ adapter->stats.xmitcalled = tx_stats.xmit_called;
+ adapter->stats.xmitfinished = tx_stats.xmit_finished;
+ adapter->stats.txbytes = tx_stats.tx_bytes;
 }
 
 static u64 *qlcnic_fill_tx_queue_stats(u64 *data, void *stats)
-- 
1.7.1